Codechange: support string temporaries as DParam/Remove: stredup

rubidium42 opened the pull request:
## Motivation / Problem

Having to keep local variables around to prevent string temporaries with the string parameter system.
The existence of `stredup`/manual memory management, especially to validate whether a game string parameter is valid.


## Description

Add a `std::unique_ptr<std::string>` to retain a copy of a string in the `StringParameter`. Not directly a `std::string`, as we need a mechanism to know whether the value is used and given temporaries are rarely used the overhead of `std::unique_ptr` over `std::optional` is negligable compared to the memory usage.

Use this new support for temporaries to not have to `stredup` string parameters, and to not have to maintain an array of fields to `free` later.

Remove `stredup`.

Use this new support for temporaries to remove the local variables in callers of `SetDParamStr`.


## Limitations

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

rubidium42 committed:
Codechange: allow string temporaries in a StringParameter

rubidium42 committed:
Codechange: use the string temporaries for game script strings

rubidium42 committed:
Remove: now unused stredup

rubidium42 committed:
Codechange: do not keep local variable for temporary string parameters

PeterN commented on src/strings.cpp:
```diff
@@ -363,6 +363,18 @@ void SetDParamStr(size_t n, const std::string &str)
+ */
+void SetDParamStr(size_t n, std::string &&str)
+{
+	_global_string_params.SetParam(n, std::move(str));
```
Is it valid to use std::move here? str might not be a temporary.

rubidium42 replied:
https://godbolt.org/z/nhr8d74ch

If there is only a `std::string &&`, then passing non-rvalues will yield a compilation error... unless you explicitly use `std::move`.
If there is only a `const std::string &`, then thing will just be as the were before.
With this change there is a `const std::string &` and `std::string &&`, which means that for rvalues the latter is called which does the `std::move` to make it implicit that it should be passed on. The former remains as is.

So, the only way to pass a `std::string` that should not be copied to `SetDParamStr` is when it's either a (temporary) rvalue, or when you explicitly `std::move` it. It's practically the same construct that is used for copy vs move constructors/assignments.

glx22 commented on src/strings.cpp:
```diff
@@ -986,26 +994,14 @@ static void FormatString(StringBuilder &builder, const char *str_arg, StringPara
-						sub_args_need_free[i] = true;
-						sub_args.SetParam(i++, (uint64)(size_t)g);
-						_game_script_raw_strings.push_back(g);
+						sub_args.SetParam(i++, std::string(s, p - s - 1)); // also skip the trailing \".
```
~~suggestion~~
						~~sub_args.SetParam(i++, std::string(s, p - s)); // also skip the trailing \".~~
~~Last char vs size I think.~~

Actually I was wrong, I built the PR to check.

glx22 commented on src/script/api/script_text.cpp:
```diff
@@ -248,8 +248,6 @@ void ScriptText::_GetEncodedText(std::back_insert_iterator<std::string> &output,
-	const std::string &encoded_text = this->GetEncodedText();
-
-	::SetDParamStr(0, encoded_text);
+	::SetDParamStr(0, this->GetEncodedText());
```
For yet unknown reason, MSVC uses `void SetDParamStr(size_t n, const std::string &str)` for this call.

End of SetDParamStr call 
![image](https://github.com/OpenTTD/OpenTTD/assets/2952192/c966cdc8-1ee6-4af6-8cbe-cd97da8d1b67)

And just before GetString call 
![image](https://github.com/OpenTTD/OpenTTD/assets/2952192/72ebed5f-f8a3-4b7b-bd8b-ed6662fb4cb0)

rubidium42 replied:
`GetEncodedText` returned `const std::string`, which means MSVC does not see it as a temporary/rvalue. By making `GetEncodedText` just return `std::string`
Note that returning `const std::string` was common practice before C++11, however since it has become less common as it hinders move semantics. Also, the language has become more strict what must be done with respect to copy elision. This copy elision seems to have been one of the reasons to return const objects.

rubidium42 committed:
Codechange: use std::string without const as return

Otherwise some compilers, e.g. MSVC, do not pick up that these are temporaries
and as such it will pass the temporaries to `const std::string &` instead of
the wanted `std::string &&`
